### ISSUE TITLE
Unable to build project because of no existing artifact

### ISSUE THREAD

**Issue by remkolodder** [privacyidea/keycloak-provider#6]
Hi,

I am trying to build the keycloak provider but it fails because the example provider 4.3.0-Final does not exist. Is there a way to resolve this?

```
$ mvn clean install wildfly:deploy
[INFO] Scanning for projects...
Downloading from central: https://repo.maven.apache.org/maven2/org/keycloak/keycloak-examples-providers-parent/4.3.0.Final/keycloak-examples-providers-parent-4.3.0.Final.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.keycloak:authenticator-required-action-example:[unknown-version]: Could not find artifact org.keycloak:keycloak-examples-providers-parent:pom:4.3.0.Final in central (https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 25, column 13
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.keycloak:authenticator-required-action-example:[unknown-version] (/home/keycloak/keycloak-6.0.1/examples/providers/keycloak-provider/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.keycloak:authenticator-required-action-example:[unknown-version]: Could not find artifact org.keycloak:keycloak-examples-providers-parent:pom:4.3.0.Final in central (https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 25, column 13 -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException

```

**Comment by Mipronimo** [comment]
Hi,
we used the Keycloak example provider 4.3.0-Final to build this project. You can find it here: https://www.keycloak.org/archive/downloads-4.3.0.html
I think, it is not the best way, but the documentation for server development is based on this demo.
Clone this repo to `IdeaProjects/keycloak-demo-4.3.0.Final/examples/providers/privacyIDEA` and you should be able to build our provider.

**Comment by remkolodder** [comment]
Ah yes, that seemed to work so far. It seems that the JAR file is at least accepted. It gives a nullpointerexception for some reason though:

13:18:03,081 WARN  [org.keycloak.services] (default task-18) KC-SERVICES0013: Failed authentication: java.lang.NullPointerException

Is there information I can provide to see what I did wrong?

**Comment by Mipronimo** [comment]
Can you please send the complete output for this warning?

**Comment by remkolodder** [comment]
Sure I can;

08:59:48,143 WARN  [org.keycloak.services] (default task-33) KC-SERVICES0013: Failed authentication: java.lang.NullPointerException
	at org.privacyidea.authenticator.privacyIDEAAuthenticator.authenticate(privacyIDEAAuthenticator.java:95)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:221)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:117)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:97)
	at org.keycloak.authentication.AuthenticationProcessor.authenticationAction(AuthenticationProcessor.java:873)
	at org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:292)
	at org.keycloak.services.resources.LoginActionsService.processAuthentication(LoginActionsService.java:263)
	at org.keycloak.services.resources.LoginActionsService.authenticate(LoginActionsService.java:259)
	at org.keycloak.services.resources.LoginActionsService.authenticateForm(LoginActionsService.java:320)
	at sun.reflect.GeneratedMethodAccessor635.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:139)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:510)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:400)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$0(ResourceMethodInvoker.java:364)
	at org.jboss.resteasy.core.interception.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:355)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:366)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:338)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:137)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:100)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:439)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:229)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:135)
	at org.jboss.resteasy.core.interception.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:355)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:138)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:215)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:227)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:56)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.keycloak.services.filters.KeycloakSessionServletFilter.doFilter(KeycloakSessionServletFilter.java:90)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.lang.Thread.run(Thread.java:748)

It results in the following error for authentication, disabling the PrivacyIDEA object in the flow resolves the login error though (so it is in that part of the service):

07:20:38,799 WARN  [org.keycloak.events] (default task-28) type=LOGIN_ERROR, realmId=<realmname>, clientId=account, userId=null, ipAddress=127.0.0.1, error=invalid_user_credentials, auth_method=openid-connect, auth_type=code, redirect_uri=https://ourhostname/auth/realms/<realmname>/account/login-redirect, code_id=83bca044-fb01-4eca-abd7-c0cc15dc54b3, username=<myuid>

**Comment by Mipronimo** [comment]
Okay thanks. I figured it out now.
We have no default value for the config verify ssl.
So if you do not change the value, it is null. I'll fix it in the next commit.
For now you can just turn the verification on and off again, and save it.
It won't be null anymore and should work.

Thanks for your report!

**Comment by remkolodder** [comment]
Hi,

Thanks for that!

I have toggled the button but it does not seem to change anything (Yet) ;-)

**Comment by remkolodder** [comment]
After moving it around a bit it now does something (I enabled and disabled both other checkboxes as well).

The attempt now gets at the privacyidea server:

[2019-05-16 13:40:38,697][7399][140595051525888][WARNING][privacyidea.lib.resolvers.LDAPIdResolver:323] failed to check password for 'e533f2d8-917a-4089-976e-cc52c68c849a'/u'myAD-DN-string': LDAPPasswordIsMandatoryError('password is mandatory in simple bind',)

**Comment by cornelinux** [comment]
This looks like you have an error in your resolver configuration.
Please provide more information about your privacyIDEA installation. At least the version number!
And you need to post your LDAP config.
This probably has nothing to do with the java project not building.

**Comment by cornelinux** [comment]
@Mipronimo so basically we can close this issue, since it is fixed, right?

**Comment by remkolodder** [comment]
Shall I put that on the community forum Cornelius? I can screenshot most of my configuration except the obvious things that we do not want to share with the world.. :)

**Comment by Mipronimo** [comment]
> @Mipronimo so basically we can close this issue, since it is fixed, right?

Yes. It is fixed in #8

**Comment by cornelinux** [comment]
@remkolodder This is most probalby a privacyidea server issue. (Well I think it is configuration, not acutally a problem in privacyidea ,-)
So either put it in the privacyIDEA github or into the community forum. Thanks!
